src/Simplex/FileTransfer/Protocol.hs

```diff
@@ -1 +1,364 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -fno-warn-unticked-promoted-constructors #-}
+
 module Simplex.FileTransfer.Protocol where
+import Data.Aeson (FromJSON (..), ToJSON (..), (.=))
+import qualified Data.Aeson as J
+import qualified Data.Aeson.Encoding as JE
+import qualified Data.Attoparsec.ByteString.Char8 as A
+import Data.ByteString.Char8 (ByteString)
+import qualified Data.ByteString.Char8 as B
+import Data.Kind
+import Data.Maybe (isNothing, isJust)
+import Data.Text.Encoding (decodeLatin1, encodeUtf8)
+import Data.Type.Equality
+import Data.Word (Word16)
+import Database.SQLite.Simple.FromField (FromField (..))
+import Database.SQLite.Simple.ToField (ToField (..))
+import Simplex.Messaging.Agent.Protocol (updateSMPServerHosts)
+import qualified Simplex.Messaging.Crypto as C
+import Simplex.Messaging.Encoding
+import Simplex.Messaging.Encoding.String
+import Simplex.Messaging.Notifications.Transport (ntfClientHandshake)
+import Simplex.Messaging.Parsers (fromTextField_)
+import Simplex.Messaging.Protocol hiding (Command (..), CommandTag (..), Recipient, Sender, SRecipient, SSender, Cmd)
+import Simplex.Messaging.Util (eitherToMaybe, (<$?>))
+import Data.Type.Equality (TestEquality (testEquality))
+import Data.Kind (Type)
+import Data.ByteString (ByteString)
+
+-- | File protocol clients
+data FileParty = Recipient | Sender
+  deriving (Show)
+
+data SFileParty :: FileParty -> Type where
+  SRecipient :: SFileParty Recipient
+  SSender :: SFileParty Sender
+
+instance TestEquality SFileParty where
+  testEquality SRecipient SRecipient = Just Refl
+  testEquality SSender SSender = Just Refl
+  testEquality _ _ = Nothing
+
+deriving instance Show (SFileParty p)
+
+class FilePartyI (p :: FileParty) where sFileParty :: SFileParty p
+
+instance FilePartyI Recipient where sFileParty = SRecipient
+
+instance FilePartyI Sender where sFileParty = SSender
+
+data FileCommandTag (p :: FileParty) where
+  NEW_ :: FileCommandTag Sender
+  ADD_ :: FileCommandTag Sender
+  UP_ :: FileCommandTag Sender
+  SDEL_ :: FileCommandTag Sender
+  DOWN_ :: FileCommandTag Recipient
+  RDEL_ :: FileCommandTag Recipient
```

Comment:
```suggestion
  FACK_ :: FileCommandTag Recipient
```

This one doesn't really delete anything, but it kind of similar to ACK in SMP.

Reply:
The server would still invalidate the address after FACK, similarly to how it deletes the stored message in SMP.